output/pixel_pusher.c

```diff
@@ -0,0 +1,272 @@
+        device->base.ui_name = output_config.pixel_pusher_grids[i].ui_name;
+        device->strip_num = output_config.pixel_pusher_grids[i].strip_num;
+
+        // Geometry and pixel arrangement
+        // TODO: Maybe do some sanity checking on configuration
+        // TODO: Maybe read this from the device itself (via the PixelPusher struct in packet.py),
+        // although because we obviously have to set the position manually in the config file
+        // it's maybe cleaner to keep width and height there as well
+        device->width = output_config.pixel_pusher_grids[i].width;
+        device->height = output_config.pixel_pusher_grids[i].height;
+        device->base.pixels.length = device->width * device->height;
+        device->base.vertex_head = output_config.pixel_pusher_grids[i].vertexlist;
+
+        int res = output_device_arrange_grid(&device->base, device->width, device->height);
+        if (res < 0) {
+            ERROR("Unable to arrange pixels for PixelPusher grid %zu", i);
+            device->base.active = false;
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+static int pp_init_out() {
+    seq_num = 0;
+
+    // Try to open a socket
+    if (out_fd > 0) {
+        close(out_fd);
+    }
+    out_fd = socket(AF_INET, SOCK_DGRAM, 0);
+    if (out_fd < 0) {
+        PERROR("Error opening PixelPusher data socket");
+        return -1;
+    }
+
+    // Target address
+    // This is only called if pp_find_pusher finishes successfully, so
+    // pp_info is set
+    out_addr = calloc(1, sizeof *out_addr);
+    out_addr->sin_family = AF_INET;
+    out_addr->sin_addr.s_addr = pp_info->header.ip_addr;
+    out_addr->sin_port = htons(pp_info->info.my_port);
+
+    out_packet = calloc(1, 4 + 2*(1+3*240));
+    if (out_packet == NULL) MEMFAIL();
+
+    return 0;
+}
+
+int output_pp_init() {
+    if (pp_find_pusher() < 0) {
+        return -1;
+    }
+
+    if (pp_init_out() < 0) {
+        return -1;
+    }
+
+    return pp_add_grids();
+}
+
+void output_pp_term() {
+    INFO("Terminating PixelPusher");
+
+    if (pp_info) {
+        free(pp_info);
+        pp_info = NULL;
+    }
+
+    for (size_t i = 0; i < n_grid_devices; i++) {
+        struct output_device base = grid_devices[i].base;
+
+        free(base.pixels.xs);
+        free(base.pixels.ys);
+        free(base.pixels.colors);
+
+        if (base.prev != NULL)
+            base.prev->next = base.next;
+        if (base.next != NULL)
+            base.next->prev = base.prev;
+    }
+    free(grid_devices);
+    grid_devices = NULL;
+    n_grid_devices = 0;
+
+
+    if (out_fd) {
+        close(out_fd);
+        out_fd = -1;
+    }
+
+    if (out_addr) {
+        free(out_addr);
+        out_addr = NULL;
+    }
+
+    if (out_packet) {
+        free(out_packet);
+        out_packet = NULL;
+    }
+}
+
+int output_pp_do_frame() {
+    seq_num++;
+
+    ((uint32_t *) out_packet)[0] = seq_num;
+    size_t out_packet_idx = 4;
```

Comment:
Magic 4 is magic; perhaps a struct pp_output_packet { uint32 seq_num; uint8_t strip_num; uint8_t payload[...]; } would be cleaner?